### PR TITLE
Add NumpyFileWrapper to numpy loader

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- Refactoring to detach numpy specific logic/members from the general FileLoader

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Created a separate NumpyFileWrapper, to hold the members that are only relevant for NumpyLoader*
     *Created a generic NumpyReader template, to have a common SetupImpl implementation for CPU and GPU versions (coming in the following PR)*

 - Affected modules and functionalities:
     *NumpyReader*
 - Key points relevant for the review:
     *NA*
 - Validation and testing:
     *Existing tests*
 - Documentation (including examples):
     *NA*

**JIRA TASK**: *[DALI-2127]*

### CODE REVIEW

File: dali/operators/reader/loader/numpy_loader.h [hunk 1]
```diff
@@ -54,6 +54,36 @@ class NumpyParseTarget{
...
+    image.set_type(type_info);
+  }
+
+  const TensorShape<>& shape() const {
+    return image.shape();
+  }
+
+  void set_shape(const TensorShape<>& shape) {
+    image.Resize(shape);
+  }
```
Comment (mzient) [inline dali/operators/reader/loader/numpy_loader.h:76]: This smells of inefficiency - a single call to `Resize(shape, type)` results in fewer reallocations.

Comment (jantonguirao) [reply dali/operators/reader/loader/numpy_loader.h]: removed, it was not used anyway

**Review by mzient — APPROVED** [review]

**Review by mzient — APPROVED** [review]

File: dali/operators/reader/loader/file_loader.h [hunk 2]
```diff
@@ -34,11 +34,21 @@
 
 namespace dali {
 
+namespace detail {
+
+template <typename T, typename result = decltype(std::declval<T &>().image)>
```
Comment (JanuszL) [inline dali/operators/reader/loader/file_loader.h:39]:
```suggestion
template <typename T, typename result = decltype(std::declval<T &>().data)>
```
?

File: dali/operators/reader/loader/file_loader.h [hunk 3]
```diff
@@ -34,11 +34,21 @@
...
+inline std::true_type HasImage(T *);
+inline std::false_type HasImage(...);
+
+/// @brief Inherits `true_type`, `if T::image` exists
+template <typename T>
+struct has_image : decltype(HasImage((T*)0)) {};  // NOLINT
```
Comment (JanuszL) [inline dali/operators/reader/loader/file_loader.h:45]:
```suggestion
struct has_image : decltype(HasData((T*)0)) {};  // NOLINT
```

Comment (JanuszL) [reply dali/operators/reader/loader/file_loader.h:45]: I don't think it is an image anymore.

File: dali/operators/reader/loader/numpy_loader.h [hunk 4]
```diff
@@ -54,6 +54,24 @@ class NumpyParseTarget{
   }
 };
 
+struct NumpyFileWrapper {
+  Tensor<CPUBackend> image;
```
Comment (JanuszL) [inline dali/operators/reader/loader/numpy_loader.h:58]:
```suggestion
  Tensor<CPUBackend> data;
```
I don't think we can call it an image anymore.

**Review by JanuszL — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Comment by jantonguirao** [comment]
!build

**Pull request merged**